pkg/services/pluginsintegration/angulardetectorsprovider/dynamic.go

```diff
@@ -0,0 +1,302 @@
+	}()
+
+	opCtx, canc := context.WithTimeout(ctx, gcomFetchTimeout)
+	defer canc()
+
+	// Fetch new patterns from GCOM, store response in db and get the corresponding detectors
+	newDetectors, err := d.fetchAndStoreDetectors(opCtx)
+	if err != nil {
+		d.log.Error("error while updating patterns", "error", err)
+		return
+	}
+
+	// Update cached detectors
+	d.setDetectors(newDetectors)
+}
+
+// backgroundJob is the function executed periodically in the background by the background service.
+// It calls tryUpdateDetectors.
+func (d *Dynamic) runBackgroundJob(ctx context.Context) {
+	d.tryUpdateDetectors(ctx)
+}
+
+// setDetectorsFromCache sets the in-memory detectors from the patterns in the store.
+func (d *Dynamic) setDetectorsFromCache(ctx context.Context) error {
+	var cachedPatterns GCOMPatterns
+	rawCached, err := d.store.Get(ctx)
+	switch {
+	case errors.Is(err, angularpatternsstore.ErrNoCachedValue):
+		// Swallow ErrNoCachedValue without changing cache
+		return nil
+	case err == nil:
+		// Try to unmarshal, convert to detectors and set local cache
+		if err := json.Unmarshal([]byte(rawCached), &cachedPatterns); err != nil {
+			return fmt.Errorf("json unmarshal: %w", err)
+		}
+		cachedDetectors, err := d.patternsToDetectors(cachedPatterns)
+		if err != nil {
+			return fmt.Errorf("convert to detectors: %w", err)
+		}
+		d.setDetectors(cachedDetectors)
+		return nil
+	default:
+		// Other error
+		return fmt.Errorf("get cached value: %w", err)
+	}
+}
+
+// Run is the function implementing the background service and updates the detectors periodically.
+func (d *Dynamic) Run(ctx context.Context) error {
+	d.log.Debug("Started background service")
+
+	// Determine when next run is, and check if we should run immediately
+	lastUpdate, err := d.store.GetLastUpdated(ctx)
+	if err != nil {
+		return fmt.Errorf("get last updated: %w", err)
+	}
+	nextRunUntil := time.Until(lastUpdate.Add(d.backgroundJobInterval))
+	if nextRunUntil <= 0 {
+		// Do first run immediately
+		d.backgroundJob.runBackgroundJob(ctx)
+		nextRunUntil = d.backgroundJobInterval
+	}
+
+	// Keep running periodically
+	ticker := time.NewTicker(nextRunUntil)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			d.backgroundJob.runBackgroundJob(ctx)
+			// Restore default TTL if we run with a shorter interval the first time
+			ticker.Reset(d.backgroundJobInterval)
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// ProvideDetectors returns the cached detectors. It returns an empty slice if there's no value.
+func (d *Dynamic) ProvideDetectors(_ context.Context) []angulardetector.AngularDetector {
+	// Wait for initial restore to be done.
+	// This channel is closed after the restore is done, so it will always return immediately once the
+	// initial restore is done.
+	<-d.initialRestore
+
+	d.mux.RLock()
+	r := d.detectors
+	d.mux.RUnlock()
+	return r
+}
+
+// Same configuration as pkg/plugins/repo/client.go
+func makeHttpClient() http.Client {
```

Comment:
Duplicated code from the signing keys fetcher. This should probably be shared with the signing keys service but I'll leave it for a follow-up PR.

Reply:
:+1: yep that should work! I'll try that, thanks

Reply:
It is also used to save the fetched value in the store afterwards, so I think it makes sense to keep it, but maybe rename it:

```go
gcomFetchTimeout    = time.Minute * 1
```

:point_down: 

```go
backgroundJobTimeout    = time.Minute * 1
```

WDYT?

Reply:
> It is also used to save the fetched value in the store afterwards

Same here about the timeout, I don't think we need to worry about that. Only about doing http requests over the internet.